src/main/java/signup_and_login/UserDBAccess.java

```diff
@@ -0,0 +1,103 @@
+     */
+    @Override
+    public List<UserModel> loadUsers() {
+        try {
+            String result = performHttpRequest("GET", null);
+
+            // Parse JSON data
+            Gson gson = new Gson();
+            JsonObject jsonObject = gson.fromJson(result, JsonObject.class);
+            JsonArray usersArray = jsonObject.get("users").getAsJsonArray();
+
+            List<UserModel> users = new ArrayList<>();
+            for (JsonElement userElement : usersArray) {
+                JsonObject userObject = userElement.getAsJsonObject();
+                String username = userObject.get("username").getAsString();
+                String password = userObject.get("password").getAsString();
+                users.add(new UserModel(username, password));
+            }
+
+            return users;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return new ArrayList<>();
+    }
+
+    /**
+     * Saves a new user to the database (effectively signing them up).
+     * @param request the UserModel object representing the new user
+     */
+    @Override
+    public void saveNewUser(UserModel request) {
+        try {
+            JsonObject userObject = new JsonObject();
+            userObject.addProperty("username", request.getUsername());
+            userObject.addProperty("password", request.getPassword());
+
+            JsonObject mainObject = new JsonObject();
+            mainObject.add("user", userObject);
+
+            String jsonInputString = mainObject.toString();
+            performHttpRequest("POST", jsonInputString);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     Performs an HTTP request to the specified URL with the given method and JSON input.
+     This method is used internally by the UserDBAccess class.
+     @param method the HTTP method (e.g., GET, POST)
+     @param jsonInputString the JSON input string for POST requests, null for GET requests
+     @return the response from the HTTP request
+     @throws Exception if an error occurs during the HTTP request
+     **/
+    private String performHttpRequest(String method, String jsonInputString) throws Exception {
+        URL url = new URL(UserDBAccess.API_URL);
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+        conn.setRequestMethod(method);
+
+        if (method.equals("POST")) {
+            conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
+            conn.setRequestProperty("Accept", "application/json");
+            conn.setDoOutput(true);
+            try (OutputStream os = conn.getOutputStream()) {
+                byte[] input = jsonInputString.getBytes(StandardCharsets.UTF_8);
+                os.write(input, 0, input.length);
+            }
+        }
+
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
```

Comment:
you should check the http response code before try and log an error otherwise

Reply:
see commit da4da39d5eb19e14f37cc1b5e3f03bbb7f81022b